### PR TITLE
Optimize one-phase commit

### PR DESCRIPTION
**Pull request by xiong-gang**
We assign distributed transaction ID for every transaction: 1PC transaction,
    2PC transaction, and implicit auto-commit transaction. The consequence is we
    have to clear the transaction ID (MyTmGxact->gxid) when the transaction ends
    and currently we need acquire ProcArrayLock to clear it, which is a huge lock
    contention on the master running OLTP workloads. The reason for taking
    ProcArrayLock is we need 'strict serialization of commits and rollbacks with
    snapshot-taking: we do not allow any transaction to exit the set of running
    transactions while a snapshot is being taken'.

    The truth is we don't actually need to consider one-phase commit transactions
    when taking a distributed snapshot. In this commit, we introduce a flag
    'MyTmGxact->includeInSnapshot' to specify should the snapshot-taking process
    peek this transaction's gxid. The flag is set right before dispatch 'PREPARE'
    command (bypass one-phase commit), and it is cleared with ProcArrayLock held to
    make sure the serialization with snapshot-taking.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/cdb/cdbtm.c [hunk 1]
```diff
@@ -841,8 +848,9 @@ prepareDtxTransaction(void)
 		 DtxStateToString(MyTmGxactLocal->state));
 
 	Assert(MyTmGxactLocal->state == DTX_STATE_ACTIVE_DISTRIBUTED);
-	Assert(MyTmGxact->gxid > FirstDistributedTransactionId);
 
+	if (MyTmGxact->gxid == InvalidDistributedTransactionId)
```
Comment (asimrp) [inline src/backend/cdb/cdbtm.c]: This condition is hard to understand for me.  If `MyTmGxact->gxid` is a valid distributed XID, why should this transaction be not included in snapshot?

File: src/backend/cdb/cdbtm.c [hunk 2]
```diff
@@ -1327,15 +1335,14 @@ dispatchDtxCommand(const char *cmd)
...
 	{
 		/* As QE doesn't take distributed snapshot, we don't have to hold ProcArrayLock to clear gxid */
 		Assert(LWLockHeldByMe(ProcArrayLock));
-		MyTmGxact->gxid = InvalidDistributedTransactionId;
+		MyTmGxact->includeInSnapshot = false;
 	}
-	else if (Gp_role == GP_ROLE_EXECUTE)
-		MyTmGxact->gxid = InvalidDistributedTransactionId;
 
+	MyTmGxact->gxid = InvalidDistributedTransactionId;
```
Comment (asimrp) [inline src/backend/cdb/cdbtm.c]: Would it make sense to have this assert here:
```C
    AssertImply(MyTmGxact->invludeInSnapshot, Gp_role == GP_ROLE_DISPATCH);
```

File: src/backend/cdb/cdbtm.c [hunk 3]
```diff
@@ -1529,64 +1528,21 @@ isDtxQueryDispatcher(void)
 }
 
 /*
- * Called prior to handling a requested that comes to the QD, or a utility request to a QE.
+ * Called prior to handling a requested that comes to the QD
```
Comment (asimrp) [inline src/backend/cdb/cdbtm.c]: s/requested/request/

File: src/backend/cdb/cdbtm.c (hunk 1 above)
Comment (xiong-gang) [reply src/backend/cdb/cdbtm.c]: I did find something wrong with this approach, I am working on another way of doing it. Thanks.

**Comment by xiong-gang** [comment]
I am going to close this PR as the optimization I was thinking is not working.

I have tried several different ways to optimize the single row
insert/update/delete to get better throughput, but I realized that there is no
easy way to do it. Here is something I learned.
Let's take a look at this case:
```
session 1: insert into t values(1);  --transaction tx1
session 1: insert into t values(2);  --transaction tx2
session 2: select * from t;          --transaction tx3
```

The valid result of tx3 is <empty> or <1> or <1,2>, it can't be <2> because tx1
is logically ahead of tx2. So tx3 must be aware of that tuple (1) on one
segment is inserted earlier than the tuple (2) on another segment, and tx3 can
get this knowledge only from the master node. The current way of doing this is
is to create a distributed transaction for both tx1 and tx2, and tx3 can get a
distributed snapshot of them.

Let's assume tx1 and tx2 don't create the distributed transaction and tx3 use
local snapshot to determine the visibility, the example transactions execute in
such order:
```
tx3 get a local snapshot on segment 1
tx1 execute on segment 1
tx2 execute on segment 2
tx3 get a local snapshot on segment 2
```
In this case, tx3 will get an invalid result <2>.


The next question is, can we make an exception for this kind of single row
insert/update/delete distributed transaction and clear the distributed
transaction id without holding ProcArrayLock? The answer is no, the reason is
the same as upstream. Let's take a look at this case:
```
session 1: begin; -- transaction tx1
session 1: update t set b=100 where a=1;
session 2: update t set b=200 where a=1; -- transaction tx2, block by tx1
session 3: select * from t;  -- take a distributed snapshot and find out tx1 is in progress
session 1: end;
session 2: tx2 is unblocked and finished
session 3: still taking the distributed snapshot and find tx2 is finished
```

In this case, session 3 get an inconsistently distributed snapshot and will see
two versions of the same tuple.

**Comment by asimrp** [comment]
👍 thank you for the detailed explanation with examples.

**Comment by ashwinstar** [comment]
@xiong-gang Do we have those cases/examples as tests in GPDB currently? If not it's worth adding them explicitly even if this PR is not going in.

**Comment by xiong-gang** [comment]
> @xiong-gang Do we have those cases/examples as tests in GPDB currently? If not it's worth adding them explicitly even if this PR is not going in.

No, I don't think we have the test cases yet. I will try to add the test in the other PR `https://github.com/greenplum-db/gpdb/pull/8826`

**Pull request closed without merging**